control-plane/agents/core/src/pool/specs.rs

```diff
@@ -414,69 +413,104 @@ impl ResourceSpecsLocked {
         specs.replicas.to_vec()
     }
 
+    /// Worker that reconciles dirty PoolSpec's with the persistent store.
+    /// This is useful when pool operations are performed but we fail to
+    /// update the spec with the persistent store.
+    pub(crate) async fn reconcile_dirty_pools(&self, registry: &Registry) -> bool {
+        let mut pending_count = 0;
+
+        let pools = self.get_locked_pools();
+        for pool_spec in pools {
+            if let Ok(_guard) = pool_spec.operation_guard(OperationMode::ReconcileStart) {
+                let creating = pool_spec.lock().status.creating();
+                if creating {
+                    if SpecOperations::delete_spec(registry, &pool_spec)
+                        .await
+                        .is_err()
+                    {
+                        pending_count += 1;
+                    }
+                    continue;
+                }
+            }
+        }
+        pending_count > 0
+    }
+
     /// Worker that reconciles dirty ReplicaSpec's with the persistent store.
     /// This is useful when replica operations are performed but we fail to
     /// update the spec with the persistent store.
     pub(crate) async fn reconcile_dirty_replicas(&self, registry: &Registry) -> bool {
-        if registry.store_online().await {
-            let mut pending_count = 0;
-
-            let replicas = self.get_replicas();
-            for replica_spec in replicas {
-                let (mut replica_clone, _guard) = {
-                    if let Ok(guard) = replica_spec.operation_guard(OperationMode::ReconcileStart) {
-                        let replica = replica_spec.lock();
-                        if !replica.status.created() {
+        let mut pending_count = 0;
+
+        let replicas = self.get_replicas();
+        for replica_spec in replicas {
+            let (mut replica_clone, _guard) = {
+                if let Ok(guard) = replica_spec.operation_guard(OperationMode::ReconcileStart) {
+                    let spec_status = replica_spec.lock().status();
+                    match spec_status {
+                        ReplicaSpecStatus::Creating | ReplicaSpecStatus::Deleted => {
+                            SpecOperations::delete_spec(registry, &replica_spec)
+                                .await
+                                .ok();
+                            continue;
+                        }
+                        ReplicaSpecStatus::Created(_) => {
+                            if replica_spec.lock().pending_op() {
+                                (replica_spec.lock().clone(), guard)
+                            } else {
+                                continue;
+                            }
+                        }
+                        ReplicaSpecStatus::Deleting => {
+                            // handled by the garbage collector
```

Comment:
This match pattern seems to be the same for multiple types of specs (i.e. nexus and pool). I wonder if we can abstract it to a common function so that the behaviour is generic across types.

Reply:
Yeah, I think we can make the whole process of reconciling dirty specs generic.